PCB format update.

sezergueler opened the pull request:
ASMs can contain extensions now. Routing Policy extension was added for testing.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/netsec-ethz/scion/1068)
<!-- Reviewable:end -->

pszalach commented:
Review status: 0 of 6 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[proto/asm_exts.capnp, line 7 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KjliCB2x-l2-bWPb7a4:-KjliCB3AWG-mJUbhQL9:b-qbi6bx) ([raw file](https://github.com/netsec-ethz/scion/blob/f7dc19da928077ef7bf62ef14deaa1455ae774e6/proto/asm_exts.capnp#L7)):*
> ```
> 
> struct RoutingPolicyExt{
>     extType @0 :UInt8;  # The extension type
> ```

1) I'm not sure do we need `extType`. Isn't type defined (implicitly) by the CapnProto definition?
2) length field is probably required for signing it securely

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1068)*
<!-- Sent from Reviewable.io -->

sezergueler commented:
Review status: 0 of 8 files reviewed at latest revision, 1 unresolved discussion.

---

*[proto/asm_exts.capnp, line 7 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KjliCB2x-l2-bWPb7a4:-KjqGNP34O5S4moJmuWy:bn01rn0) ([raw file](https://github.com/netsec-ethz/scion/blob/f7dc19da928077ef7bf62ef14deaa1455ae774e6/proto/asm_exts.capnp#L7)):*
<details><summary><i>Previously, pszalach (Paweł Szałachowski) wrote…</i></summary><blockquote>

1) I'm not sure do we need `extType`. Isn't type defined (implicitly) by the CapnProto definition?
2) length field is probably required for signing it securely

</blockquote></details>

Yes, you are right. The type is defined implicitly by CapnProto. I removed this field. 
However, I do not see against what attack the length field would be a countermeasure. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1068)*
<!-- Sent from Reviewable.io -->

pszalach commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 8 files reviewed at latest revision, 1 unresolved discussion.

---

*[proto/asm_exts.capnp, line 7 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KjliCB2x-l2-bWPb7a4:-KjqKl9knvCZuthZME-t:b-tzpg0b) ([raw file](https://github.com/netsec-ethz/scion/blob/f7dc19da928077ef7bf62ef14deaa1455ae774e6/proto/asm_exts.capnp#L7)):*
<details><summary><i>Previously, sezergueler wrote…</i></summary><blockquote>

Yes, you are right. The type is defined implicitly by CapnProto. I removed this field. 
However, I do not see against what attack the length field would be a countermeasure. 
</blockquote></details>

Ok, let's discuss it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-:-KjqL7EViEV3hXFq58Ma:bnfp4nl)*
<!-- Sent from Reviewable.io -->

pszalach commented:
Reviewed 2 of 6 files at r1, 6 of 6 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1068)*
<!-- Sent from Reviewable.io -->

kormat commented:
Reviewed 2 of 6 files at r1, 3 of 6 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[infrastructure/beacon_server/base.py, line 304 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkB-vvg2YJ5RosG_54I:-KkB-vvg2YJ5RosG_54J:bc4gzdc) ([raw file](https://github.com/netsec-ethz/scion/blob/e54429f85a3bded2e736ab19c5bdace897a00492/infrastructure/beacon_server/base.py#L304)):*
> ```Python
>             logging.debug("%s", pcb.sibra_ext)
>         for asm in pcb.iter_asms():
>             if asm.p.exts.policy:
> ```

This is _always_ true. It does not actually check for the presence of the extension.

---

*[infrastructure/beacon_server/base.py, line 327 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkB0j1jvNkUeMzknNLp:-KkB0j1jvNkUeMzknNLq:bb53502) ([raw file](https://github.com/netsec-ethz/scion/blob/e54429f85a3bded2e736ab19c5bdace897a00492/infrastructure/beacon_server/base.py#L327)):*
> ```Python
>         chain = self._get_my_cert()
>         _, cert_ver = chain.get_leaf_isd_as_ver()
>         test_isd_ases = [ISD_AS.from_values(1, 10), ISD_AS.from_values(2, 23)]
> ```

Uh, this would completely break the test network if it were enforced. This does not seem like a good idea.

---

*[infrastructure/path_server/base.py, line 378 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkB0w1LGG7eLaoV7nVB:-KkB0w1LGG7eLaoV7nVC:b-aq9pqn) ([raw file](https://github.com/netsec-ethz/scion/blob/e54429f85a3bded2e736ab19c5bdace897a00492/infrastructure/path_server/base.py#L378)):*
> ```Python
>             logging.debug("%s", pcb.sibra_ext)
>         for asm in pcb.iter_asms():
>             if asm.p.exts.policy:
> ```

This is also _always_ true.

---

*[lib/packet/asm_exts.py, line 27 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkAzxj99qtrpoBEQyx4:-KkAzxj99qtrpoBEQyx5:b-ppfjqc) ([raw file](https://github.com/netsec-ethz/scion/blob/e54429f85a3bded2e736ab19c5bdace897a00492/lib/packet/asm_exts.py#L27)):*
> ```Python
> 
> class RoutingPolicyExt(Cerealizable):
>     NAME = "PolicyExt"
> ```

`RoutingPolicyExt`

---

*[lib/packet/asm_exts.py, line 47 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkB-CmmmMEjKGIlEoNf:-KkB-CmmmMEjKGIlEoNg:bligrzk) ([raw file](https://github.com/netsec-ethz/scion/blob/e54429f85a3bded2e736ab19c5bdace897a00492/lib/packet/asm_exts.py#L47)):*
> ```Python
>         if self.VER != 2:
>             raise SCIONSigVerError(
>                 "RoutingPolicyExt.sig_pack3 cannot support version %s", self.VER)
> ```

`sig_pack2`

---

*[lib/packet/asm_exts.py, line 56 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkB-mAo4aIxgbstfAZI:-KkB-mApilFNy-AaOgWK:bglex4k) ([raw file](https://github.com/netsec-ethz/scion/blob/e54429f85a3bded2e736ab19c5bdace897a00492/lib/packet/asm_exts.py#L56)):*
> ```Python
>     def short_desc(self):
>         a = []
>         a.append("RoutingPolicyExt extension: Policy type: %s, Interface: %s" %
> ```

"ASes:" at the end.

---

*[lib/packet/asm_exts.py, line 57 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkB-YENauih-VYjOPLR:-KkB-YEOEFxNnTz5dD8w:b-azkobd) ([raw file](https://github.com/netsec-ethz/scion/blob/e54429f85a3bded2e736ab19c5bdace897a00492/lib/packet/asm_exts.py#L57)):*
> ```Python
>         a = []
>         a.append("RoutingPolicyExt extension: Policy type: %s, Interface: %s" %
>                  (self.p.polType, self.p.itf))
> ```

`RoutingPolType.to_str(self.p.polType)`

---

*[lib/packet/asm_exts.py, line 59 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkB-i2bJFgPYupOcxae:-KkB-i2bJFgPYupOcxaf:brz7xgk) ([raw file](https://github.com/netsec-ethz/scion/blob/e54429f85a3bded2e736ab19c5bdace897a00492/lib/packet/asm_exts.py#L59)):*
> ```Python
>                  (self.p.polType, self.p.itf))
>         for isd_as in self.p.isdases:
>             a.append(" %s" % isd_as)
> ```

`ISD_AS(isd_as)`

---

*[lib/packet/pcb.py, line 97 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkAzsjXQOqwbs1zg5eW:-KkAzsjXQOqwbs1zg5eX:bt5e5c1) ([raw file](https://github.com/netsec-ethz/scion/blob/e54429f85a3bded2e736ab19c5bdace897a00492/lib/packet/pcb.py#L97)):*
> ```Python
> 
>     @classmethod
>     def from_values(cls, isd_as, trc_ver, cert_ver, pcbms, hashTreeRoot, mtu, exts,
> ```

I would make `exts` an optional parameter. `exts=()`.

---

*[lib/packet/pcb.py, line 121 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkAz7ZbcOtrsVfRfW0k:-KkAz7Zc4FbpS28F6Mhi:b4b1rzh) ([raw file](https://github.com/netsec-ethz/scion/blob/e54429f85a3bded2e736ab19c5bdace897a00492/lib/packet/pcb.py#L121)):*
> ```Python
> 
>     def routing_pol_ext(self):
>         return RoutingPolicyExt(self.p.exts.policy)
> ```

What does this do if there is no routing policy on the ASMarking (and how is that detected?)?

---

*[lib/packet/pcb.py, line 147 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkAzaSYT8_6Jw2HDX6U:-KkAzaSYT8_6Jw2HDX6V:b-h3j8w) ([raw file](https://github.com/netsec-ethz/scion/blob/e54429f85a3bded2e736ab19c5bdace897a00492/lib/packet/pcb.py#L147)):*
> ```Python
>         b.append(self.p.hashTreeRoot)
>         b.append(self.p.mtu.to_bytes(2, 'big'))
>         b.append(self.routing_pol_ext().sig_pack2())
> ```

This makes the routing policy extension mandatory for all PCBs?

---

*[proto/asm_exts.capnp, line 8 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkAy3YrTHbT9ZzmEGqj:-KkAy3YrTHbT9ZzmEGqk:b-rs652g) ([raw file](https://github.com/netsec-ethz/scion/blob/e54429f85a3bded2e736ab19c5bdace897a00492/proto/asm_exts.capnp#L8)):*
> ```
> struct RoutingPolicyExt{
>     polType @0 :UInt8;  # The policy type
>     itf @1 :UInt64;
> ```

This name is very opaque. I had to look at the python code to figure out what it's for. `ifID` is the name used in the other capnp definitions.

---

*[proto/pcb.capnp, line 28 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkAxRHhL-L-VMJJ9LRZ:-KkAxRHhL-L-VMJJ9LR_:bjtdg5n) ([raw file](https://github.com/netsec-ethz/scion/blob/e54429f85a3bded2e736ab19c5bdace897a00492/proto/pcb.capnp#L28)):*
> ```
>     mtu @7 :UInt16;  # Internal MTU
>     exts :group {
>         policy @8 :Exts.RoutingPolicyExt;
> ```

i would name this `routingPolicy`, as we may want to add other policy extensions later.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1068)*
<!-- Sent from Reviewable.io -->

sezergueler commented:
Review status: 2 of 10 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[infrastructure/beacon_server/base.py, line 304 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkB-vvg2YJ5RosG_54I:-KkjV38D_6XWOsgxxinT:b-mamr74) ([raw file](https://github.com/netsec-ethz/scion/blob/e54429f85a3bded2e736ab19c5bdace897a00492/infrastructure/beacon_server/base.py#L304)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

This is _always_ true. It does not actually check for the presence of the extension.
</blockquote></details>

How can we check the presence?  

---

*[infrastructure/beacon_server/base.py, line 327 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkB0j1jvNkUeMzknNLp:-KkjVFhxCbpnWhttAmZ-:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/e54429f85a3bded2e736ab19c5bdace897a00492/infrastructure/beacon_server/base.py#L327)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

Uh, this would completely break the test network if it were enforced. This does not seem like a good idea.
</blockquote></details>

Done.

---

*[lib/packet/asm_exts.py, line 27 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkAzxj99qtrpoBEQyx4:-KkjVHs6QBRYZVTx46XL:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/e54429f85a3bded2e736ab19c5bdace897a00492/lib/packet/asm_exts.py#L27)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

`RoutingPolicyExt`
</blockquote></details>

Done.

---

*[lib/packet/asm_exts.py, line 47 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkB-CmmmMEjKGIlEoNf:-KkjVIxqr1SR6283ziM_:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/e54429f85a3bded2e736ab19c5bdace897a00492/lib/packet/asm_exts.py#L47)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

`sig_pack2`
</blockquote></details>

Done.

---

*[lib/packet/asm_exts.py, line 56 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkB-mAo4aIxgbstfAZI:-KkjVKYaPqCXkHF0GC7N:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/e54429f85a3bded2e736ab19c5bdace897a00492/lib/packet/asm_exts.py#L56)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

"ASes:" at the end.
</blockquote></details>

Done.

---

*[lib/packet/asm_exts.py, line 57 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkB-YENauih-VYjOPLR:-KkjVLTXmuHnyociL5qu:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/e54429f85a3bded2e736ab19c5bdace897a00492/lib/packet/asm_exts.py#L57)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

`RoutingPolType.to_str(self.p.polType)`
</blockquote></details>

Done.

---

*[lib/packet/asm_exts.py, line 59 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkB-i2bJFgPYupOcxae:-KkjVNhDIcKUndjvRkQf:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/e54429f85a3bded2e736ab19c5bdace897a00492/lib/packet/asm_exts.py#L59)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

`ISD_AS(isd_as)`
</blockquote></details>

Done.

---

*[lib/packet/pcb.py, line 97 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkAzsjXQOqwbs1zg5eW:-KkjVPCk77X6ifPthJm_:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/e54429f85a3bded2e736ab19c5bdace897a00492/lib/packet/pcb.py#L97)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

I would make `exts` an optional parameter. `exts=()`.
</blockquote></details>

Done.

---

*[proto/asm_exts.capnp, line 7 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KjliCB2x-l2-bWPb7a4:-KkjVVKVI1smcIIqdZXL:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/f7dc19da928077ef7bf62ef14deaa1455ae774e6/proto/asm_exts.capnp#L7)):*
<details><summary><i>Previously, pszalach (Paweł Szałachowski) wrote…</i></summary><blockquote>

Ok, let's discuss it.
</blockquote></details>

Done.

---

*[proto/asm_exts.capnp, line 8 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkAy3YrTHbT9ZzmEGqj:-KkjVWMvSf-3LC4yCNof:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/e54429f85a3bded2e736ab19c5bdace897a00492/proto/asm_exts.capnp#L8)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

This name is very opaque. I had to look at the python code to figure out what it's for. `ifID` is the name used in the other capnp definitions.
</blockquote></details>

Done.

---

*[proto/pcb.capnp, line 28 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkAxRHhL-L-VMJJ9LRZ:-KkjVY4Fk-TZG6UaVCMR:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/e54429f85a3bded2e736ab19c5bdace897a00492/proto/pcb.capnp#L28)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

i would name this `routingPolicy`, as we may want to add other policy extensions later.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1068)*
<!-- Sent from Reviewable.io -->

kormat commented:
Reviewed 1 of 8 files at r4.
Review status: 3 of 10 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[infrastructure/beacon_server/base.py, line 304 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkB-vvg2YJ5RosG_54I:-KkoSjdcAuLpmDgbHapI:bfmj6b7) ([raw file](https://github.com/netsec-ethz/scion/blob/e54429f85a3bded2e736ab19c5bdace897a00492/infrastructure/beacon_server/base.py#L304)):*
<details><summary><i>Previously, sezergueler wrote…</i></summary><blockquote>

How can we check the presence?  
</blockquote></details>

With the current proto definition, you need to be able to distinguish a zero-valued `routingPolicy` from one which is populated.

The alternative is to change how the proto definition works, and to use a list of unions for the extensions. Something like:

```
struct ASExt {
  union {
    routingPolicy @0 :Exts.RoutingPolicyExt;
  }
}

struct ASMarking {
  ...
  exts @8 :List(ASExt);
}
```
That way you can determine the number and types of extensions present by examining `exts`.

---

*[sub/lwip, line 1 at r4](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkoRiPITuG5GsF4YAhj:-KkoRiPITuG5GsF4YAhk:bwjsv67) ([raw file](https://github.com/netsec-ethz/scion/blob/4d59e537e7773a26b0cf81657f2e7011957d2a70/sub/lwip#L1)):*
> ```
> Subproject commit 06acc9e1d61eff87a0a7026e82d57b20a8b107aa
> ```

You do not want to do this. Same for `lwip-contrib`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1068)*
<!-- Sent from Reviewable.io -->

kormat commented:
Reviewed 2 of 7 files at r5.
Review status: 3 of 10 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[lib/packet/pcb.py, line 97 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkAzsjXQOqwbs1zg5eW:-KkuJWY4YLLAQ0FVenFo:byljol7) ([raw file](https://github.com/netsec-ethz/scion/blob/e54429f85a3bded2e736ab19c5bdace897a00492/lib/packet/pcb.py#L97)):*
<details><summary><i>Previously, sezergueler wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Never use a mutable initial value for an optional argument (which is why i suggested a tuple). http://python-guide-pt-br.readthedocs.io/en/latest/writing/gotchas/#mutable-default-arguments

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1068)*
<!-- Sent from Reviewable.io -->

sezergueler commented:
Review status: 1 of 10 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[infrastructure/beacon_server/base.py, line 304 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkB-vvg2YJ5RosG_54I:-KkyeL7HzZGKAOPDf8Em:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/e54429f85a3bded2e736ab19c5bdace897a00492/infrastructure/beacon_server/base.py#L304)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

With the current proto definition, you need to be able to distinguish a zero-valued `routingPolicy` from one which is populated.

The alternative is to change how the proto definition works, and to use a list of unions for the extensions. Something like:

```
struct ASExt {
  union {
    routingPolicy @0 :Exts.RoutingPolicyExt;
  }
}

struct ASMarking {
  ...
  exts @8 :List(ASExt);
}
```
That way you can determine the number and types of extensions present by examining `exts`.
</blockquote></details>

Done.

---

*[infrastructure/path_server/base.py, line 378 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkB0w1LGG7eLaoV7nVB:-KkyeNQu2xYsA_J8Q24N:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/e54429f85a3bded2e736ab19c5bdace897a00492/infrastructure/path_server/base.py#L378)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

This is also _always_ true.
</blockquote></details>

Done.

---

*[lib/packet/pcb.py, line 97 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkAzsjXQOqwbs1zg5eW:-KkyeOk7FSPY7mzFLIIB:b-b7y428) ([raw file](https://github.com/netsec-ethz/scion/blob/e54429f85a3bded2e736ab19c5bdace897a00492/lib/packet/pcb.py#L97)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

Never use a mutable initial value for an optional argument (which is why i suggested a tuple). http://python-guide-pt-br.readthedocs.io/en/latest/writing/gotchas/#mutable-default-arguments
</blockquote></details>

Done. Thank you!!

---

*[lib/packet/pcb.py, line 121 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkAz7ZbcOtrsVfRfW0k:-KkyeT16Ficq52lGWlxZ:b-bqlb9g) ([raw file](https://github.com/netsec-ethz/scion/blob/e54429f85a3bded2e736ab19c5bdace897a00492/lib/packet/pcb.py#L121)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

What does this do if there is no routing policy on the ASMarking (and how is that detected?)?
</blockquote></details>

This does not get called if there are no extensions. 

---

*[lib/packet/pcb.py, line 147 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkAzaSYT8_6Jw2HDX6U:-Kkyey4_3zRYB7LxtNkv:b-elwvdg) ([raw file](https://github.com/netsec-ethz/scion/blob/e54429f85a3bded2e736ab19c5bdace897a00492/lib/packet/pcb.py#L147)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

This makes the routing policy extension mandatory for all PCBs?
</blockquote></details>

We do not want to make any extension mandatory. If there are no extensions, we do not enter the loop, right?

---

*[sub/lwip, line 1 at r4](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkoRiPITuG5GsF4YAhj:-Kkyejo47rwaoXEBIQIh:b1ck40c) ([raw file](https://github.com/netsec-ethz/scion/blob/4d59e537e7773a26b0cf81657f2e7011957d2a70/sub/lwip#L1)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

You do not want to do this. Same for `lwip-contrib`.
</blockquote></details>

I will fix this when rebasing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1068)*
<!-- Sent from Reviewable.io -->

kormat commented:
Reviewed 1 of 4 files at r6.
Review status: 2 of 10 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[lib/packet/pcb.py, line 124 at r6](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KlO57_Au8P8ZsHECWDI:-KlO57_Au8P8ZsHECWDJ:b-g9vgd4) ([raw file](https://github.com/netsec-ethz/scion/blob/52882918cac63e392529e3ed60c61f4401579fa8/lib/packet/pcb.py#L124)):*
> ```Python
>             yield self.asm_ext(ext)
> 
>     def asm_ext(self, ext_raw):
> ```

It looks this is only called from a simple place, and it's not clear to me that it's useful to call it from elsewhere either, so i suggest just making `iter_exts` do `yield ASMExt(ext)`, and remove this method.

---

*[lib/packet/pcb.py, line 152 at r6](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KlO5arcZkr9Zph3vW7q:-KlO5arcZkr9Zph3vW7r:bvqs7ms) ([raw file](https://github.com/netsec-ethz/scion/blob/52882918cac63e392529e3ed60c61f4401579fa8/lib/packet/pcb.py#L152)):*
> ```Python
>         b.append(self.p.mtu.to_bytes(2, 'big'))
>         for ext in self.iter_exts():
>             b.append(ext.sig_pack0())
> ```

Ah. We have a versioning problem here. `sig_pack0` assumes that all extensions only have a single field, and that this never changes. `ASMarking` needs to know the version...

(This is where i spiralled off into thought, before realising that my advice to use a list of extensions was a mistake. Sorry!)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1068)*
<!-- Sent from Reviewable.io -->

sezergueler commented:
Review status: 0 of 11 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[lib/packet/pcb.py, line 124 at r6](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KlO57_Au8P8ZsHECWDI:-KlT9Ky2J6VsxQCvXWt4:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/52882918cac63e392529e3ed60c61f4401579fa8/lib/packet/pcb.py#L124)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

It looks this is only called from a simple place, and it's not clear to me that it's useful to call it from elsewhere either, so i suggest just making `iter_exts` do `yield ASMExt(ext)`, and remove this method.
</blockquote></details>

Done.

---

*[lib/packet/pcb.py, line 152 at r6](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KlO5arcZkr9Zph3vW7q:-KlT9MZw5I0L9V-T5tuL:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/52882918cac63e392529e3ed60c61f4401579fa8/lib/packet/pcb.py#L152)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

Ah. We have a versioning problem here. `sig_pack0` assumes that all extensions only have a single field, and that this never changes. `ASMarking` needs to know the version...

(This is where i spiralled off into thought, before realising that my advice to use a list of extensions was a mistake. Sorry!)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1068)*
<!-- Sent from Reviewable.io -->

kormat commented:
Reviewed 7 of 9 files at r7.
Review status: 7 of 11 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[infrastructure/beacon_server/base.py, line 327 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkB0j1jvNkUeMzknNLp:-KlTEGrLR_533bkoVZro:b-te1io3) ([raw file](https://github.com/netsec-ethz/scion/blob/e54429f85a3bded2e736ab19c5bdace897a00492/infrastructure/beacon_server/base.py#L327)):*
<details><summary><i>Previously, sezergueler wrote…</i></summary><blockquote>

Done.
</blockquote></details>

This still will break the test network if enforced. You can't add a routing policy extension at all by default.

---

*[infrastructure/beacon_server/base.py, line 318 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KlTE60lGHQe5OafI3-s:-KlTE60mDACC0Uz4U8OE:bk55eih) ([raw file](https://github.com/netsec-ethz/scion/blob/d3fbe16e571721ef40ba3c968e524b9b0a694a5c/infrastructure/beacon_server/base.py#L318)):*
> ```Python
>         raise NotImplementedError
> 
>     def _create_routing_pol_ext(self, pol_type, interface, isd_ases):
> ```

I'm not seeing the value in this method, if all it does is pass the values directly to `RoutingPolicyExt.from_values`, and is only called once in any case.

---

*[infrastructure/path_server/base.py, line 378 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KlTDlUMAtBADbXcOM8D:-KlTDlUMAtBADbXcOM8E:b-3l57is) ([raw file](https://github.com/netsec-ethz/scion/blob/d3fbe16e571721ef40ba3c968e524b9b0a694a5c/infrastructure/path_server/base.py#L378)):*
> ```Python
>             logging.debug("%s", pcb.sibra_ext)
>         for asm in pcb.iter_asms():
>             if asm.p.exts.routingPolicy.set:
> ```

This would change to:
```
pol = asm.routing_pol_ext()
if pol:
  self.handle_routing_pol_ext(pol)
```

---

*[lib/packet/asm_exts.py, line 38 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KlTBp5_4PpQbmGQ0u86:-KlTBp5_4PpQbmGQ0u87:b-k96s0z) ([raw file](https://github.com/netsec-ethz/scion/blob/d3fbe16e571721ef40ba3c968e524b9b0a694a5c/lib/packet/asm_exts.py#L38)):*
> ```Python
>         for i, isd_as in enumerate(isd_ases):
>             p.isdases[i] = int(isd_as)
>         cls.extType = ASMExtType.ROUTING_POLICY
> ```

This creates/modifies a _class_ attribute. Class attributes should never be modified, and should always be initialised in the top-level of the class. Add `EXT_TYPE = ASMExtType.ROUTING_POLICY` just below the `NAME` class attribute.

---

*[lib/packet/asm_exts.py, line 50 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KlTDEaC_8Of63TcE8EX:-KlTDEaC_8Of63TcE8EY:b9ecywl) ([raw file](https://github.com/netsec-ethz/scion/blob/d3fbe16e571721ef40ba3c968e524b9b0a694a5c/lib/packet/asm_exts.py#L50)):*
> ```Python
>         b.append(self.p.set.to_bytes(1, 'big'))
>         b.append(self.p.polType.to_bytes(1, 'big'))
>         b.append(self.p.ifID.to_bytes(2, 'big'))
> ```

This is too small. `ifID` is a `UInt64`.

---

*[lib/packet/asm_exts.py, line 55 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KlTDPcLVvLSaEo2s9oU:-KlTDPcLVvLSaEo2s9oV:b-nwjfuz) ([raw file](https://github.com/netsec-ethz/scion/blob/d3fbe16e571721ef40ba3c968e524b9b0a694a5c/lib/packet/asm_exts.py#L55)):*
> ```Python
>         return b"".join(b)
> 
>     def sig_pack0(self):
> ```

I don't think it makes sense to have this.

---

*[lib/packet/pcb.py, line 123 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KlTAB3YmcAwOfub01aL:-KlTAB3YmcAwOfub01aM:b-vjkjpf) ([raw file](https://github.com/netsec-ethz/scion/blob/d3fbe16e571721ef40ba3c968e524b9b0a694a5c/lib/packet/pcb.py#L123)):*
> ```Python
> 
>     def routing_pol_ext_set(self):
>         return self.p.exts.routingPolicy.set
> ```

I don't think this method is necessary. I would probably just change `routing_pol_ext` to:
```
if self.p.exts.routingPolicy.set:
  return RoutingPolicyExt(self.p.exts.routingPolicy)
return None
```

---

*[lib/packet/pcb.py, line 133 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KlTAWoLlm4_nLECSXg3:-KlTAWoLlm4_nLECSXg4:b1vwiy5) ([raw file](https://github.com/netsec-ethz/scion/blob/d3fbe16e571721ef40ba3c968e524b9b0a694a5c/lib/packet/pcb.py#L133)):*
> ```Python
>         self.p.from_dict(d)
> 
>     def sig_pack8(self):
> ```

There are now 9 fields in the ASMarking proto definition. It probably doesn't make sense to add an extension that isn't implemented, as there's no way to include it in verification/signing.

---

*[lib/packet/pcb.py, line 151 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KlTArFnFu6M4d7Bcsi5:-KlTArFnFu6M4d7Bcsi6:bdorwr5) ([raw file](https://github.com/netsec-ethz/scion/blob/d3fbe16e571721ef40ba3c968e524b9b0a694a5c/lib/packet/pcb.py#L151)):*
> ```Python
>             b.append(self.routing_pol_ext().sig_pack3())
>         else:
>             b.append(self.routing_pol_ext().sigpack0())
> ```

Why include an empty extension?

---

*[proto/asm_exts.capnp, line 15 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KlT9NXMnhOKzG3ptQpe:-KlT9NXNGgHuW45dVZAY:bys4ets) ([raw file](https://github.com/netsec-ethz/scion/blob/d3fbe16e571721ef40ba3c968e524b9b0a694a5c/proto/asm_exts.capnp#L15)):*
> ```
> struct ISDAnnouncementExt{
>     set @0 :Bool;   # TODO(Sezer): Implement announcement extension
> }
> ```

Newline missing at end of file.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1068)*
<!-- Sent from Reviewable.io -->

sezergueler commented:
Review status: 1 of 11 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[infrastructure/beacon_server/base.py, line 327 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KkB0j1jvNkUeMzknNLp:-KlTW_lEFdK9CxYykfju:b-qqu3cu) ([raw file](https://github.com/netsec-ethz/scion/blob/e54429f85a3bded2e736ab19c5bdace897a00492/infrastructure/beacon_server/base.py#L327)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

This still will break the test network if enforced. You can't add a routing policy extension at all by default.
</blockquote></details>

Ok, I removed it. It was to test that extensions are added and received correctly.

---

*[infrastructure/beacon_server/base.py, line 318 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KlTE60lGHQe5OafI3-s:-KlTWYtZKR-jgsuFOfep:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/d3fbe16e571721ef40ba3c968e524b9b0a694a5c/infrastructure/beacon_server/base.py#L318)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

I'm not seeing the value in this method, if all it does is pass the values directly to `RoutingPolicyExt.from_values`, and is only called once in any case.
</blockquote></details>

Done.

---

*[infrastructure/path_server/base.py, line 378 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KlTDlUMAtBADbXcOM8D:-KlTWo4C6FknVltdj8vF:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/d3fbe16e571721ef40ba3c968e524b9b0a694a5c/infrastructure/path_server/base.py#L378)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

This would change to:
```
pol = asm.routing_pol_ext()
if pol:
  self.handle_routing_pol_ext(pol)
```
</blockquote></details>

Done.

---

*[lib/packet/asm_exts.py, line 38 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KlTBp5_4PpQbmGQ0u86:-KlTWpZ-pSGqfh_q9JM3:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/d3fbe16e571721ef40ba3c968e524b9b0a694a5c/lib/packet/asm_exts.py#L38)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

This creates/modifies a _class_ attribute. Class attributes should never be modified, and should always be initialised in the top-level of the class. Add `EXT_TYPE = ASMExtType.ROUTING_POLICY` just below the `NAME` class attribute.
</blockquote></details>

Done.

---

*[lib/packet/asm_exts.py, line 50 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KlTDEaC_8Of63TcE8EX:-KlTWqCf3XHCsAmRO38g:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/d3fbe16e571721ef40ba3c968e524b9b0a694a5c/lib/packet/asm_exts.py#L50)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

This is too small. `ifID` is a `UInt64`.
</blockquote></details>

Done.

---

*[lib/packet/asm_exts.py, line 55 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KlTDPcLVvLSaEo2s9oU:-KlTWs4F1zA8-68z7mQy:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/d3fbe16e571721ef40ba3c968e524b9b0a694a5c/lib/packet/asm_exts.py#L55)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

I don't think it makes sense to have this.
</blockquote></details>

Done.

---

*[lib/packet/pcb.py, line 123 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KlTAB3YmcAwOfub01aL:-KlTWuALS2z1Usrrkmah:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/d3fbe16e571721ef40ba3c968e524b9b0a694a5c/lib/packet/pcb.py#L123)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

I don't think this method is necessary. I would probably just change `routing_pol_ext` to:
```
if self.p.exts.routingPolicy.set:
  return RoutingPolicyExt(self.p.exts.routingPolicy)
return None
```

</blockquote></details>

Done.

---

*[lib/packet/pcb.py, line 133 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KlTAWoLlm4_nLECSXg3:-KlTXuD8X9GGpVpwjTWo:b-9jm656) ([raw file](https://github.com/netsec-ethz/scion/blob/d3fbe16e571721ef40ba3c968e524b9b0a694a5c/lib/packet/pcb.py#L133)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

There are now 9 fields in the ASMarking proto definition. It probably doesn't make sense to add an extension that isn't implemented, as there's no way to include it in verification/signing.
</blockquote></details>

Done. I removed the isd announcement extension from as marking.

---

*[lib/packet/pcb.py, line 151 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KlTArFnFu6M4d7Bcsi5:-KlTWxIchcd5pGDRFH8l:by9ait1) ([raw file](https://github.com/netsec-ethz/scion/blob/d3fbe16e571721ef40ba3c968e524b9b0a694a5c/lib/packet/pcb.py#L151)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

Why include an empty extension?
</blockquote></details>

Done. I misunderstood it.

---

*[proto/asm_exts.capnp, line 15 at r7](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KlT9NXMnhOKzG3ptQpe:-KlTWzZ7cBh86rkd2zPz:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/d3fbe16e571721ef40ba3c968e524b9b0a694a5c/proto/asm_exts.capnp#L15)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

Newline missing at end of file.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1068)*
<!-- Sent from Reviewable.io -->

kormat commented:
Please rebase and fix the git submodules. Apart from that, one minor comment, and then this PR should be ready for merging.

---

Reviewed 6 of 6 files at r8, 1 of 1 files at r9, 1 of 1 files at r10.
Review status: 8 of 11 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[test/lib/packet/pcb_test.py, line 67 at r10](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KlXoKyyyWSa6_Z8g2kD:-KlXoKyzTIYR5q5jjeu5:b-d9yek0) ([raw file](https://github.com/netsec-ethz/scion/blob/96a3d2703e0ae677f52eb3094a7d754af7cd00b7/test/lib/packet/pcb_test.py#L67)):*
> ```Python
>             pcbms.append(create_mock_full({"p": "pcbm %d" % i}))
>         exts = []
>         exts.append(create_mock_full({"EXT_TYPE": 0, "p":
> ```

Use the constant from `lib.types` here instead of `0`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-:-KlXoY0xJcC5pp4-38Zr:b-af12p7)*
<!-- Sent from Reviewable.io -->

sezergueler commented:
Review status: 7 of 11 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[test/lib/packet/pcb_test.py, line 67 at r10](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-KlXoKyyyWSa6_Z8g2kD:-KlXtugToL_0I0y6MCsm:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/96a3d2703e0ae677f52eb3094a7d754af7cd00b7/test/lib/packet/pcb_test.py#L67)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

Use the constant from `lib.types` here instead of `0`.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1068)*
<!-- Sent from Reviewable.io -->

kormat commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r11, 4 of 5 files at r12.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1068#-:-KlYALfF1vpvEBnvOtI3:bnfp4nl)*
<!-- Sent from Reviewable.io -->
